Project Walking/Source code/app/src/main/java/com/teamapricot/projectwalking/handlers/StorageHandler.java

```diff
@@ -37,12 +38,52 @@
 
     private final AppCompatActivity activity;
 
+    private static StorageHandler instance;
+
+    private PhotoDao photoDao = null;
+    private RouteDao routeDao = null;
+
+    private File lastSavedPhoto = null;
     /**
-     * Creates a new instance of the {@code ImageFileHandler} class, using the specified {@code AppCompatActivity}
-     * @param activity The activity used for retrieving external storage directory paths
+     * Private constructor.
+     * @param activity The associated activity
      */
-    public StorageHandler(AppCompatActivity activity) {
+    private StorageHandler(AppCompatActivity activity) {
         this.activity = activity;
+        Database.getDatabase(activity.getApplicationContext()).thenAccept(database -> {
+            photoDao = database.photoDao();
+            routeDao = database.routeDao();
+        });
```

Comment:
Maybe (I'm not sure myself) it would make sense here to do something like

```java
Database database = Database.getDatabase(activity.getApplicationContext()).get();
photoDao = database.photoDao();
routeDao = database.routeDao();
```

**Pros:** You'll get rid of the methods `databaseIsReady` and `waitForDatabase`.
**Cons:** The `.get()` blocks the current thread while retrieving the `Database`, more or less acting like a synchronous call instead of an asynchronous call.